### ISSUE TITLE
如何使用模型内部的inputs_embeds，插入新的embedding

### ISSUE THREAD

**Issue by SZUHvern** [QwenLM/Qwen2.5-VL#1315]
非常感谢您之前的帮助！我已经解决了很多问题。
我希望能在Qwen2.5-VL的基础框架上作出改动。思路是在原始messages中加入额外的占位符，在embeding层将这个占位符替换为其它模型的输出特征。例如文本中新增"<new prompt>",且tokenizer等其它操作都已经跟随新占位符更新。我的代码如下：

inputs_embeds = model.model.embed_tokens(inputs["input_ids"])
pixel_values = inputs["pixel_values"].type(model.visual.dtype)
image_embeds = model.visual(pixel_values, grid_thw=inputs["image_grid_thw"])
n_image_tokens = (inputs["input_ids"] == model.config.image_token_id).sum().item()
n_image_features = image_embeds.shape[0]
if n_image_tokens != n_image_features:
    raise ValueError(
        f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
    )

# 获取原版inputs_embeds
mask = inputs["input_ids"] == model.config.image_token_id
mask_unsqueezed = mask.unsqueeze(-1)
mask_expanded = mask_unsqueezed.expand_as(inputs_embeds)
image_mask = mask_expanded.to(inputs_embeds.device)
image_embeds = image_embeds.to(inputs_embeds.device, inputs_embeds.dtype)
inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)

# 获取新加入的embeds
prompt_embs = self.seg_fc(embeddings)

# 将其插入自定义prompt_token_idx的位置，其它操作模仿原版获取inputs_embeds的流程
mask = inputs["input_ids"] == self.config.prompt_token_idx
mask_unsqueezed = mask.unsqueeze(-1)
mask_expanded = mask_unsqueezed.expand_as(inputs_embeds)
image_mask = mask_expanded.to(inputs_embeds.device)
prompt_embs = prompt_embs.to(inputs_embeds.device, inputs_embeds.dtype)
inputs_embeds = inputs_embeds.masked_scatter(image_mask, prompt_embs)

attention_mask = inputs["attention_mask"].to(device=inputs_embeds.device)
position_ids = inputs["position_ids"].to(device=inputs_embeds.device) if "position_ids" in inputs else None

#####--------训练--------#####
lm_outputs = model(
    inputs_embeds=inputs_embeds,
    attention_mask=attention_mask,
    position_ids=position_ids,
    output_hidden_states=True,
    labels=inputs['labels'],
)
lm_loss = lm_outputs.loss
#####--------------------#####

                
#####--------inference--------#####               
gen_ids = model.generate(
    inputs_embeds=inputs_embeds,
    attention_mask=attention_mask,
    position_ids=position_ids,
    max_new_tokens=max_new_tokens,
    use_cache=use_cache,
    **generate_kwargs
)
generated_ids_trimmed = [out_ids[len(in_ids):] for in_ids, out_ids in zip(det_inputs["input_ids"], gen_ids)]
preds_prompt = processor.batch_decode(generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False)
#####--------inference--------#####      

但是我的训练曲线和结果都很差，我非常想知道是不是我的这个设计流程有问题，例如与get_rope_index_25或者attention_mask 冲突等。非常感激您的关注，您能帮忙看看有没有什么解决思路吗？

**Comment by SZUHvern** [comment]
感谢您的付出，非常期待您的回复！

**Comment by Djerry-h** [comment]
> 感谢您的付出，非常期待您的回复！

请问您解决了吗？
